src/main/java/org/elasticsearch/index/engine/Engine.java

```diff
@@ -1009,6 +1009,21 @@ public void release() {
 
     protected abstract void closeNoLock(String reason) throws ElasticsearchException;
 
+    public void flushAndClose() throws IOException {
+        if (isClosed.get() == false) {
+            logger.trace("flushAndClose now acquire writeLock");
+            try (ReleasableLock _ = writeLock.acquire()) {
+                logger.trace("flushAndClose now acquired writeLock");
+                try {
+                    logger.debug("flushing shard on close - this might take some time to sync files to disk");
+                    flush();
```

Comment:
should we think about the fact that flush might not be allowed since a recovery is happening (that will eventually get cancelled since we are closing the engine)